lms/services/digest.py

```diff
@@ -0,0 +1,267 @@
+from dataclasses import dataclass, field
+from functools import lru_cache
+from typing import List, Optional
+
+from sqlalchemy import func, select
+
+from lms.models import (
+    Assignment,
+    AssignmentGrouping,
+    AssignmentMembership,
+    Course,
+    Grouping,
+    LTIRole,
+    User,
+)
+from lms.services.h_api import HAPI
+from lms.services.mailchimp import EmailRecipient, EmailSender, MailchimpService
+
+
+class DigestService:
+    def __init__(
+        self, db, helper, h_api, mailchimp_service, sender
+    ):  # pylint:disable=too-many-arguments
+        self.db = db
+        self.helper = helper
+        self.h_api = h_api
+        self.mailchimp_service = mailchimp_service
+        self.sender = sender
+
+    def send_instructor_email_digests(
+        self, h_userids, since, until, override_to_email=None
+    ):
+        annotations = self.h_api.get_annotations(h_userids, since, until)
+        unified_courses = self.helper.unified_courses(annotations)
+
+        for h_userid in h_userids:
+            unified_user = UnifiedUser.make(self.db, h_userid)
+            digest = self.helper.instructor_digest(unified_user, unified_courses)
+
+            if not digest["num_annotations"]:
+                continue
+
+            if override_to_email is None:
+                to_email = unified_user.email
+            else:
+                to_email = override_to_email
+
+            self.mailchimp_service.send_template(
+                "instructor-email-digest",
+                self.sender,
+                recipient=EmailRecipient(to_email, unified_user.display_name),
+                template_vars=digest,
+            )
```

Comment:
So the actual public part of `DigestService` is just this one method. This is the top-level "controller" method for all the `DigestService` could below. The top-level controller algorithm is basically this:

```python
annotations = HAPI.get_annotations(*args)
unified_courses = DigestHelper.unified_courses(annotations)

for user in users:
    digest = DigestHelper.instructor_digest(user, unified_courses)
    mailchimp_service.send(digest)
```

With various minor details omitted, e.g. not sending empty digests, and supporting `override_to_email`.

So `DigestHelper` is split into these two methods `DigestHelper.unified_courses()` which is meant to be called only once and then `DigestHelper.instructor_digest(user, unified_courses)` which is meant to be called in a loop for each user. This split is the main simplification compared to my previous attempt at this: previously it tried to do it all in one method.